website: Write up documentation for Waypoint Pipelines (#3777)
hashicorp/waypoint: website/content/docs/pipelines/config.mdx

---
layout: docs
page_title: Pipelines Configuration
description: |-
  The page for detailing Waypoint pipeline configurations
---

> # Pipelines Configuration

Reviewer: Should we pluralize `pipeline(s)` or `configuration(s)`?

Author: Hm, how about `Pipeline Configurations`?

Reviewer: Sounds good!

This section details some of the various step built-in plugins and their

> configuration options.

Reviewer: I cannot parse this sentence 😅

Author: I think I flipped some words.

> This section details some of the various *built-in step plugins* and their configuration options

Maybe?

Reviewer: 👍🏻

## Built-In Step Config Options

Currently, Pipeline steps have "built-in" plugins that are backed by the existing

> Waypoint job system. Below documents each of these built-in steps and their

Reviewer: ```suggestion
Waypoint job system. See below examples of each of these built-in steps and their
```

parameters.

### `build`

This built-in Step will run a Build operation job, similar to running `waypoint build`.

```hcl
step "my-build-step" {
  use "build" {
    disable_push = false
  }
}
```

##### disable_push

- Type: **bool**
- **Optional**

### `deploy`

This built-in Step will run a Deploy operation job, similar to running `waypoint deploy`.

```hcl
step "my-deploy-step" {

>   use "deploy" {
>   }

Reviewer: ```suggestion
  use "deploy" {}
```

}
```

### `release`

This built-in step will run a Release operation job, similar to running `waypoint release`.

```hcl
step "my-release-step" {
  use "release" {
    prune        = true
    prune_retain = 4
  }
}
```

##### deployment_ref

> 0 or unset means do a release with the latest deployment. Otherwise this accepts
> a Deployment sequence to release.

Reviewer: What's a Deployment sequence? Is that the ID? Something else? Should we link elsewhere in our docs to tell people about how to find that?
```suggestion
0 or unset will use the latest deployment for a release. To specify a Deployment
for release, this accepts Deployment sequence numbers.
```

- **Optional**
- Default: 2

##### prune_retain_override

If true, will respect the `prune_retain` value when pruning during a release.

- Type: **bool**
- **Optional**

### `up`

This built-in step will run an Up operation job, similar to running `waypoint up`.

```hcl
step "my-up-step" {
  use "up" {
    prune = true
  }
}
```

##### prune

If true, will prune old releases beyond the `prune_retain` limit when a new release is created.

- Type: **bool**
- **Optional**

##### prune_retain

The total number of deployments to retain when pruning on a release. This will
only be valid if `prune_retain_override` is set to true.

- Type: **int**
- **Optional**
- Default: 2

##### prune_retain_override

If true, will respect the `prune_retain` value when pruning during a release.

- Type: **bool**
- **Optional**

### `exec`

```hcl
step "my-exec-step" {
  use "exec" {
    command = "echo"
    args    = ["hello world!"]
  }
}
```

##### command

The command to execute on the container for this Step.

- Type: **string**
- **Required**

##### args

The arguments to pass through to the command when running on the container for this Step.

- Type: **string**
- **Optional**

### `pipeline`

> This built-in step lets you reference another pipeline within a Step to be run.

Reviewer: We should consistently capitalize all `step`s

---
layout: docs
page_title: Pipelines Configuration
description: |-
  The page for detailing Waypoint pipeline configurations
---

# Pipeline Configurations

This section details some of the various built-in Step plugins and their
configuration options.

## Step Configurations

Steps have a few basic configuration options that are defined below.

##### name

```hcl
// this step has a label name "my-step-name"
step "my-step-name" {
  use "build" {}
}
```

- Stanza label
- Type: **string**
- **Required**

##### depends_on

A step can depend on many other steps within a pipeline. By default, a step will
always depend on the previously defined step unless `depends_on` is set. No

> step cycles are allowed.

Reviewer 2: "No cyclic dependencies are permitted", for example.

> step cycles are allowed.

Author: Sounds good!
